src/brogue/IO.c

```diff
@@ -3153,34 +3329,62 @@ void flavorMessage(char *msg) {
     }
 }
 
+// Insert or collapse a new message into the archive and redraw the recent
+// message display.  An incoming message may "collapse" into another (be
+// dropped in favor of bumping a repetition count) if the two have identical
+// content and one of two other conditions is met.  First, if the two messages
+// arrived on the same turn, they may collapse.  Alternately, they may collapse
+// if the older message is the latest one in the archive and the new one is not
+// semi-colon foldable (such as a combat message.)
 void messageWithoutCaps(char *msg, enum messageFlags flags) {
     short i;
+    archivedMessage *archiveEntry;
+    boolean newMessage;
+
     if (!msg[0]) {
         return;
     }
 
-    // need to confirm the oldest message? (Disabled!)
-    /*if (!messageConfirmed[MESSAGE_LINES - 1]) {
-        //refreshSideBar(-1, -1, false);
-        displayMoreSign();
-        for (i=0; i<MESSAGE_LINES; i++) {
-            messageConfirmed[i] = true;
+    // Add the message to the archive, bumping counts for recent duplicates
+    newMessage = true;
+
+    // For each at most MESSAGE_ARCHIVE_ENTRIES - 1 past entries..
+    for (i = 1; i < MESSAGE_ARCHIVE_ENTRIES; i++) {
+        archiveEntry = getArchivedMessage(i);
+
+        // Consider messages that arrived this turn for collapsing.  Also
+        // consider the latest entry (which may be from past turns) if the
+        // incoming message is not semi-colon foldable.
+        if (!((i == 1 && !(flags & FOLDABLE)) || archiveEntry->turn == rogue.playerTurnNumber)) {
```

Comment:
Why can't foldable messages collapse with the latest one?

Reply:
There's not a hard technical reason (though there is a soft one.)  The soft technical reason is that currently some function interfaces assume all folded messages occurred on the same player turn.  I could refactor around needing that assumption.

The real reason is that combat messages are the only foldable messages right now, and collapsing combat messages across turns struck me as perhaps too confusing to the player.  By breaking them up along turn lines we still get a clean delineation of rounds.

It might be better to extract a `COMBAT` value `messageType`, and set it to `FOLDABLE`, then explicitly talk about not collapsing `COMBAT` messages across turn boundaries.  Early versions of the branch did that, but it felt (slightly) like overkill and I dropped it.

Reply:
Oh I see, that makes sense, it's good as it is :+1: